comment_scan/chatter.rb

```diff
@@ -0,0 +1,122 @@
+require "chatx"
+require "htmlentities"
+
+class Chatter
+    attr_reader :HQroom, :rooms
+
+    def initialize(chatXuser, chatXpwd, hqroom, *rooms)
+        if ENV['SHORT_LOGS']
+          $stdout.sync = true #Do we really need this??
+          log_formatter = proc do |severity, datetime, progname, msg|
+            "#{msg}\n"
+          end
+        else
+          log_formatter = nil
+        end
+
+        @chatbot = ChatBot.new(chatXuser, chatXpwd, log_location: STDOUT, log_formatter: log_formatter)
+        @HQroom = hqroom.to_i
+        @rooms = rooms - [@HQroom] #Don't include HQ room in rooms
+
+        @chatbot.login(cookie_file: 'cookies.yml')
+        @chatbot.say("_Starting at rev #{`git rev-parse --short HEAD`.chop} on branch #{`git rev-parse --abbrev-ref HEAD`.chop} (#{`git log -1 --pretty=%B`.gsub("\n", '')})_", @HQroom)
+        @chatbot.join_room @HQroom
+        @chatbot.join_rooms @rooms # THIS IS THE PROBLEM
+
+        @reply_actions = Hash.new()
+        @command_actions = Hash.new()
+        @mention_actions = [] 
+        @fall_through_actions = [] 
+
+        #TODO: mention_received logic will be run alongside both command and reply logic. Going to need to fix this at some point
+        (@rooms + [@HQroom]).each do |room_id|
+            @command_actions[room_id] = Hash.new()
+
+            @chatbot.add_hook(room_id, 'message') do |message|
+                message_received(room_id, message)
+            end
+
+            @chatbot.add_hook(room_id, 'reply') do |message|
+                reply_received(room_id, message)
+                mention_received(room_id, message) #Treat replies as mentions
+            end
+
+            @chatbot.add_hook(room_id, 'mention') do |message|
+                mention_received(room_id, message)
+            end
+        end
+    end
+
+    def add_command_action(room_id, command, action, args_to_pass=nil)
+        @command_actions[room_id][command] = [action, args_to_pass]
+    end
+
```

Comment:
In the methods below this, you're kinda inconsistent about which instance variables you've initialized and which you haven't. I'd initialize them all.

Reply:
I'm not sure what you mean here. Command actions are 1 to 1. For each command, we have one action to execute, so it's a hash of command_str to [action, args] array. But reply actions are a little more difficult. Since a reply of "tp" could be to a comment, asking us to increment the tp count or a reply to some howgood output asking us to fetch tp comments. So reply_actions are a hash of reply_str to an array of [action, args] arrays. Does that make sense?

Reply:
Whaaaat?! That's some crazy black magic. I love it.